[MRG] Allow nan/inf in feature selection

adpeters opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Closes #10821
Closes #10985

#### What does this implement/fix? Explain your changes.
Allows for NaN/Inf values in `RFE/RFECV.fit` method as well as `SelectorMixin.transform`. This affects all feature selection estimators that inherit from `SelectorMixin` (except for univariate selectors), which includes those in `sklearn.feature_selection.variance_threshold` and `sklearn.linear_model.randomized_l1`. 

The `RFE/RFECV.fit` method does not need to check `y`, as any checks should be done by the estimator when it runs its own `fit`, so I changed check_X_y to check_array with just X, and allowed NaN/Inf values.

For `SelectorMixin.transform`, the method itself does not require no NaN/Inf, so we should let any inheritors to do that check themselves if they need it.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
removed check for nan/inf in SelectorMixin.tranform, RFE.fit, and RFECV.fit

Commit:
removed extra blank line

Commit:
added tests for changes to RFE/RFECV and SelectorMixin

jnothman commented:
You have test failures. Also, we are working towards a release, which this will not be included in, so please ping for review after 0.20 is released.

georgewambold commented:
Is anyone working on this issue? I'm also running into issues with `check_X_y` in RFE and I think this fix would be great.

jnothman commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -354,3 +354,34 @@ def test_rfe_cv_groups():
     )
     est_groups.fit(X, y, groups=groups)
     assert est_groups.n_features_ > 0
+
```
PEP8: there should be two blank lines here

jnothman commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -354,3 +354,34 @@ def test_rfe_cv_groups():
+    rfe = RFE(estimator=clf)
+    rfe.fit(X, y)
+    rfe.transform(X)
+
```
Two blank lines

jnothman commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -354,3 +354,34 @@ def test_rfe_cv_groups():
+    rfecv = RFECV(estimator=clf)
+    rfecv.fit(X, y)
+    rfecv.transform(X)
+
```
No blank lines

jnothman commented on sklearn/feature_selection/tests/test_rfe.py:
```diff
@@ -354,3 +354,34 @@ def test_rfe_cv_groups():
+    rfe.fit(X, y)
+    rfe.transform(X)
+
+def test_rfecv_allow_nan_inf_in_x():
```
use a loop or pytest.mark.parametrize rather than duplicating code

adpeters replied:
Added pytest.mark.parametrize.

jnothman reviewed (commented):
This is looking pretty good. Please add a test that transforming for univariate selection with NaN at transform time is acceptable. But we could (and should) also use `force_all_finite=False` in `univariate_selection._BaseFilter.fit` since the underlying scoring functions check it. `SelectFromModel` should already be lenient in fit.

You could consider creating a `feature_selection/tests/test_common.py` file that checks this for all the feature selection estimators (although we would have to use `nanvar`; `mean_variance_axis` should already exclude NaNs).

It would be really wonderful to have all feature selectors be insensitive to missing values and this is only a few lines of code away.

Commit:
pep8 changes and pytest.mark.parametrize for nan_inf tests

Commit:
added force_all_finite=False for univariate feature selection fit

Commit:
added tests for nan and inf being allowed in univariate select

Commit:
added test for nan and inf being allowed in SelectorMixin transform

Commit:
updated checks to not run check_estimators_nan_inf on feature selectors

Commit:
pep8 corrections

Commit:
created dummy_score function for use by all tests needing a fake scorer

Commit:
fixed travis test error with RFECV not having cv specified

jnothman commented:
please merge in an updated master. Circle CI should not be running on Python 2 anymore.

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -103,7 +104,7 @@ def _yield_non_meta_checks(name, estimator):
         yield check_pipeline_consistency
 
-    if name not in ALLOW_NAN:
+    if name not in ALLOW_NAN and not isinstance(estimator, SelectorMixin):
```
Hmm... Not sure if we'd rather this. There still may be selectors (not in our library) that should error on NaN/inf, and many of these selectors with their default parameters still should/will error on NaN/inf.

In the (near) future, estimator tags (#8022) will solve this.

jnothman replied:
Why do they fail the pickle test with default parameters? I think for default parameters all the feature selectors should still raise error on nan/inf except for VarianceThreshold so you should not otherwise need to modify estimator_checks should you?

adpeters replied:
If you add them to ALLOW_NAN then they fail the pickle test because they don’t allow nan/inf with the default scorer. However if you don’t add them to ALLOW_NAN then check_estimators_nan_inf errors because the transform method does not check for nan/inf.

jnothman replied:
Thanks for explaining.
I'm trying to work out if there's a better solution than creating `ALLOW_NAN_TRANSFORM`. Basically, at the moment we need to classify an estimator (and its parameter setting) as allowing or disallowing NaNs. Alternatively we could weaken the pickle test to fit on non-NaN data in the case that fitting with NaNs raises an error. (The ability to fit on NaN data there is not the point of the check. It's the interpretation of NaNs in `transform`/`predict` if NaNs are pickled with the model parameters/attributes that is the concern of that check.) Then we would be redefining ALLOW_NAN to mean "allows NaN in transform/predict, and may also allow NaN in fit". Your thoughts?

jnothman replied:
Sorry this is harder than I thought!

adpeters replied:
Thanks for helping work through it. I think relaxing the pickle test makes sense because it's not supposed to be directly checking NaN/Inf handling in the estimators. There are really two issues here: 
1) a discrepancy in whether an estimator allows NaN/Inf between its different methods (e.g. VarianceThreshold currently allows in transform but not fit), and 
2) the default parameter instance of an estimator not exhibiting its allowance of NaN/Inf. 

Your suggestion of `ALLOW_NAN_TRANSFORM` would solve 1 and I could try to implement it if we decide it's the right way to go. But 2 would still not be solved since these estimators would belong in the `ALLOW_NAN_FIT` category but still error in the pickle check with NaN. So I think weakening the pickle check is the best option without changing the default scorer. I've updated my code to include this new version of `check_estimators_pickle`.

Commit:
Merge branch 'master' into allow-nan-inf-in-feature-selection

georgewambold commented:
Thanks @jnothman for the quick comments and @adpeters for the changes, and thank you both for your time!

Commit:
allowed no NaN in pickle check for estimators in ALLOW_NAN

jnothman reviewed (commented):
I think VarianceThreshold should be updated as well. All it should require is use of `nanvar` as far as I can see (and testing in both the sparse and dense cases).

I'm still not certain that the change to the pickle test is the best thing to do, but it's okay.

Are there other changes you hope to make before making this MRG?

When the scope of the PR is finalised, please add an entry to the change log at `doc/whats_new/v0.21.rst`. Like the other entries there, please reference this pull request with `:issue:` and credit yourself (and other contributors if applicable) with `:user:`

jnothman commented:
It would probably be good to document this feature somewhere, e.g. in Notes
sections of estimator docstrings.

Commit:
added support for nan/inf in VarianceThreshold.fit, added tests for it

Commit:
added whats new documentation for feature selection allowing NaN/Inf

Commit:
added Notes for modified feature selectors about allowing NaN/Inf

Commit:
fixed pep8 error

Commit:
shortened line length for whats new documentation

adpeters commented:
Okay I updated VarianceThreshold, added tests, and added to Notes documentation and whats_new. 
I don't really like the change to the pickle test either, but I'm not sure of a better way right now. The other option would be relaxing the default scorer, `f_classif`, to allow NaN/Inf just by removing rows that contain them, but I don't think that's how we want functions to handle those cases.

I think this is all I can think of for this PR right now. Should I change it to MRG?

jnothman reviewed (commented):
Otherwise looking good!

Commit:
changed VarianceThreshold to only allow nan, not inf. Updated tests for it

Commit:
fixed phrasing in whats_new for #11635

jnothman commented on sklearn/feature_selection/tests/test_variance_threshold.py:
```diff
@@ -26,3 +27,13 @@ def test_variance_threshold():
+
+    for X in [arr, csr_matrix(arr), csc_matrix(arr), bsr_matrix(arr)]:
+        sel = VarianceThreshold().fit(X)
+        assert_array_equal([0, 3, 4], sel.get_support(indices=True))
```
I don't mind which as long as the feature is rejected. It is probably simpler to set it to 0 at fit time.

Commit:
added test for all-NaN feature in VarianceThreshold

jnothman reviewed (approved):
Nice work!

adpeters committed:
Merge branch 'master' into allow-nan-inf-in-feature-selection

adpeters committed:
Merge branch 'master' into allow-nan-inf-in-feature-selection

adpeters committed:
Merge branch 'master' into allow-nan-inf-in-feature-selection

bukson commented:
Hello, any chances to merge this into master?

jnothman commented:
I think it would be a great candidate for a quick second review, but we won't merge without it. It's quite an elegant change.

 There also appear to be merge errors.

jnothman committed:
Merge branch 'master' into HEAD

jnothman committed:
Fix use of check_array

jnothman committed:
Use check_X_y

NicolasHug commented on sklearn/feature_selection/tests/test_feature_select.py:
```diff
@@ -667,3 +671,26 @@ def test_mutual_info_regression():
+
+    univariate_filter = GenericUnivariateSelect(f_regression,
+                                                mode='percentile')
+    univariate_filter.fit(X, y)
```
we could fit with nan / inf too and remove `test_univariate_nan_inf_allowed_in_fit`

adpeters replied:
@NicolasHug the reason they were separate was that fit can only be tested with the dummy scorer because all of the actual scorers don't allow nan/inf. I feel like it's a "stronger" test to test transform with nan/inf with a real scorer in use, rather than just the dummy one. Maybe it doesn't matter though and I could combine them both to just use the dummy_score?

NicolasHug commented on sklearn/feature_selection/from_model.py:
```diff
@@ -227,3 +231,6 @@ def partial_fit(self, X, y=None, **fit_params):
         return self
+
+    def _more_tags(self):
+        return {'allow_nan': True}
```
I feel like this should be delegated to the estimator, so as to avoid the hack in `check_estimators_pickle`. WDYT @jnothman?

jnothman replied:
Yes, let's return allow_nan accoring to the base estimator.

jnothman commented on doc/whats_new/v0.21.rst:
```diff
@@ -282,6 +282,22 @@ Support for Python 3.4 and below has been officially dropped.
   passing the file name(s) or the file object(s) to the analyzer.
   :issue:`13641` by `Adrin Jalali`_.
 
+:mod:`sklearn.feature_selection`
```
Please move to v0.22.rst, thanks @adpeters

NicolasHug commented:
@adpeters please let us know if you're planning to address the remaining comments or if we should take it over.
Thanks

adpeters commented:
@NicolasHug I'll take a look at it and try to address them! Thanks for your help.

Commit:
updated SelectFromModel allow_nan tag to inherit from its estimator

Commit:
Merge branch 'master' into allow-nan-inf-in-feature-selection

Commit:
moved documentation for #11635 to v0.22

adpeters commented on sklearn/feature_selection/from_model.py:
```diff
@@ -227,3 +231,7 @@ def partial_fit(self, X, y=None, **fit_params):
         self.estimator_.partial_fit(X, y, **fit_params)
         return self
+
+    def _more_tags(self):
```
@jnothman @NicolasHug I updated this to inherit allow_nan from its estimator. This also required a change to SelectorMixin in feature_selection/base.py to satisfy estimator checks on transform. Also, I don't think this change fixes the need for the hack in check_estimators_pickle because that is necessary to handle the score_func's in univariate selectors, which aren't estimators and thus don't have the allow_nan tag. One option would be to add tags to those functions indicating whether they allow nan anyways. What do you think?

Commit:
fixed formatting

Commit:
removed test_transform_accepts_nan_inf from test_from_model.py since its not true anymore

Commit:
updated test_transform_accepts_nan_inf to handle NaN tag

Commit:
fixed formatting

Commit:
fixed formatting

jnothman commented:
I agree that it's hard to define univariate scorers' allow_nan tag without
knowing how the score function behaves. We could define it as
allow_nan=False for default params but allow_nan=True otherwise? Or
allow_nan=unknown (which is not something that currently exists)? Maybe
estimator tags demand too much of the developer when the estimator's
properties are based on complex parameters. Thoughts, @amueller?

amueller commented:
I think allowing this for RFECV and SelectFromModel is fine. We can use the tags from the estimators. For the univariate ones, we don't have tags for functions, so we can't really say. We can solve both of the mentioned issues by not changing the behavior of the univariate estimators, right?

jnothman commented:
I feel that not changing the behaviour of the univariate estimators is not
a very happy solution...

amueller commented:
@jnothman why? Is using custom score funcs that can handle NaNs a common use-case?

jnothman commented:
People with missing values want to select features too! But yes, it's a
somewhat theoretical concern

amueller commented:
Well it would be less theoretical if our functions would support missing values and skip nans (which actually would be kinda nice and easy).
But it still would only be somewhat of an edge-case.
Btw, we can't even know if a ``SelectKBest`` instance is meant to be use on classification or regression targets, which is also potentially problematic for tags.
I think a solution that's only for meta-estimators is fine for now and we can have future us worry about the function mechanisms.

jnothman commented:
I don't mind leaving this behind but I think "it doesn't fit into how we
tag" is a bad reason to reject functionality.

amueller commented:
I think what I was actually thinking of was leaving behind tests in the common tests, not leaving behind functionality. So I'd give ``SelectKBest`` more conservative tags. That would mean that someone couldn't automatically discover that ``SelectKBest`` with the right function would allow NaN, but we can still allow that and we can also still test for it (in the SelectKBest tests).

jnothman commented:
Okay, so are you proposing that we tag it to allow nan in all cases but
that the common tests do not actually check if NaN is allowed, only that
nan is disallowed when that tag is not set?

amueller commented:
currently the defaults don't allow nan, right? So I'd set the tag not not allow nan and everything passes, right?

jnothman commented:
I suppose that will work for the default score_func

lschneidpro commented:
Hi there,

I have been trying to run RFECV with lightGBM that handles NaN. 
My X has obviously some NaNs and i blocked at "check_X_y" function.
I manually commented the function but obtained other errors elsewhere.

I was wondering if someone manually fixed this issue or if there was a dev version with a fix.

I'm new to this github contribution environment so i'm not sure to understand the progresses made by this feed.

Thanks a lot for your help

NicolasHug commented:
@lschneidpro  this is an issue that this PR should solve. When it is merged, the fix would be available on the master branch, and then in the next stable release.

dmexs commented:
> People with missing values want to select features too! But yes, it's a somewhat theoretical concern

I've been using XGBoost to model lately which is happy to accept missing values and works well with the sklearn API. I had hoped to make use of SelectFromModel to perform feature reduction and keep an eye on model performance.

Looks like I'll have to implement this myself for now but look forward to this PR being merged.

jnothman commented on sklearn/feature_selection/univariate_selection.py:
```diff
@@ -361,6 +362,9 @@ def fit(self, X, y):
         pass
 
+    def _more_tags(self):
+        return {'allow_nan': True}
```
@amueller thinks we should change this to False, at least for the default score_func. Then this seems close to merge?

jnothman commented:
@adpeters, are you around to finish this off?

I'd like to see this merged as a small change that makes working with missing values potentially much smoother (people don't and shouldn't expect this to break).

adpeters committed:
set allow_nan tag to false for univariate feature selectors since the default score_funcs do not currently allow nans

adpeters committed:
merge with master

adpeters committed:
updated VarianceThreshold.fit() to handle nan when threshold is 0

adpeters committed:
removed test that no longer applies now that allow_nan tag is False for univariate selectors

adpeters committed:
added pytest import since it was accidentally removed

adpeters committed:
removed duplicate numpy import

adpeters committed:
updated assert_equals to assert

jnothman commented on sklearn/feature_selection/_base.py:
```diff
@@ -71,7 +71,9 @@ def transform(self, X):
-        X = check_array(X, dtype=None, accept_sparse='csr')
+        tags = self._get_tags()
+        X = check_array(X, dtype=None, accept_sparse='csr',
+                        force_all_finite=not tags.get('allow_nan', True))
```
this logic now won't work for univariate :(

adpeters replied:
Ah yeah I removed the test for the univariate but I wasn't sure what to do for that case. I guess the only way you could really use it would be to subclass and override the tags.

jnothman commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -361,6 +362,9 @@ def fit(self, X, y):
         pass
 
+    def _more_tags(self):
+        return {'allow_nan': False}
```
This should probably have a `FIXME` or similar to say this should depend on the score func

jnothman reviewed (commented):
Thanks for continuing on this. I wonder if we should pull univariate into a separate PR and merge the rest :|

adpeters commented:
Yeah if we're not quite sure how we want it to work, we could leave that part out of it and just leave Univariate as not allowing nan/inf until it gets figured out.

jnothman commented:
We'd like to release very soon. If you could strip this back to leave out univariate, and then propose that in another PR with a question on what to do about tags, I think that would be the most pragmatic solution.

adpeters committed:
reverted changes to univariate feature selectors to break those out to separate PR

adpeters committed:
updated documentation to reflect removal of univariate changes

adpeters committed:
fixed rfe handling of tags so they are properly inherited from the estimator

adpeters committed:
added get_tags to mockclassifier for rfe tests

jnothman reviewed (approved):
Thank you. Can we get another reviewer before release??

amueller commented:
I'm slightly confused (but also severely jet-lagged so that might be it).
The ``transform`` method of a feature selector **always** supports NaN, right? It's only ``fit`` that depends on the base estimator, right?

jnothman commented:
At master, transform does the default check for finiteness. We could make
it always permissive but not sure if that plays nicely with estimator
checks.

amueller commented:
hm... but do we want to restrict this because of estimator checks? I'm not sure if we check nan in transform. Would the "right" fix be to add separate tags for fit and transform?
I was just very confused because the test is for ``NaN`` in transform, not in fit, which seemed very strange to me.

amueller commented on sklearn/feature_selection/tests/test_from_model.py:
```diff
@@ -320,3 +336,25 @@ def test_threshold_without_refitting():
     assert X_transform.shape[1] > model.transform(data).shape[1]
+
+
+def test_transform_accepts_nan_inf():
```
can you also check for ``fit``, maybe use ``HistGradientBoostingClassifier``?

amueller reviewed (approved):
happy to go with this solution for the release, adding a test for SelectFromModel with NaN in fit would be nice, though.

adpeters committed:
added test for nan/inf in fit of SelectFromModel

adpeters committed:
added noqa tag to import to avoid flake errors

adpeters commented:
Thanks for the reviews. I added a test for NaN and Inf in `SelectFromModel.fit` using `HistGradientBoostingClassifier` as suggested. We still have to use a different classifier for the transform test since `HistGradientBoostingClassifier` does not generate feature importance metrics that `SelectFromModel` relies on to transform, but I think it's still an effective test of `fit` for `SelectFromModel`.

jnothman commented:
Thanks @adpeters! It's been very nice working with you.

adpeters commented:
@jnothman thanks for all your help on this! Glad I could contribute a little.
